crates/flux-infer/src/lean_format.rs

```diff
@@ -0,0 +1,189 @@
+                    args.iter().map(LeanExpr).format(" ")
+                )
+            }
+            Expr::And(exprs) => {
+                write!(f, "({})", exprs.iter().map(LeanExpr).format(" && "))
+            }
+            Expr::Or(exprs) => {
+                write!(f, "({})", exprs.iter().map(LeanExpr).format(" || "))
+            }
+            _ => todo!(),
+        }
+    }
+}
+
+impl<'a> fmt::Display for LeanFunDef<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let LeanFunDef(FunDef { name, args, out, comment: _, body }) = self;
+        writeln!(
+            f,
+            "def {} {} : {} :=",
+            LeanVar(name),
+            args.iter()
+                .map(|(arg, arg_sort)| format!("({} : {})", LeanVar(arg), LeanSort(arg_sort)))
+                .collect::<Vec<_>>()
+                .join(" "),
+            LeanSort(out)
+        )?;
+        writeln!(f, "  {}", LeanExpr(body))
+    }
+}
+
+impl<'a> fmt::Display for LeanPred<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            Pred::Expr(expr) => write!(f, "{}", LeanExpr(expr)),
+            Pred::And(preds) => write!(f, "({})", preds.iter().map(LeanPred).format(" ∧ ")),
+            Pred::KVar(_, _) => panic!("kvars should not appear when encoding in lean"),
+        }
+    }
+}
+
+impl<'a> fmt::Display for LeanConstraint<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            Constraint::ForAll(bind, inner) => {
+                if bind.pred.is_trivially_true() {
+                    write!(
+                        f,
+                        "(∀ ({} : {}), {})",
+                        LeanVar(&bind.name),
+                        LeanSort(&bind.sort),
+                        LeanConstraint(inner)
+                    )
+                } else {
+                    write!(
+                        f,
+                        "(∀ ({} : {}), ({} -> {}))",
+                        LeanVar(&bind.name),
+                        LeanSort(&bind.sort),
+                        LeanPred(&bind.pred),
+                        LeanConstraint(inner)
+                    )
+                }
+            }
+            Constraint::Conj(constraints) => {
+                write!(f, "({})", constraints.iter().map(LeanConstraint).format(" ∧ "))
+            }
+            Constraint::Pred(pred, _) => {
+                write!(f, "{}", LeanPred(pred))
+            }
+        }
+    }
+}
+
+impl<'genv, 'tcx> fmt::Display for LeanEncoder<'genv, 'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "-- GENERATED; DO NOT EDIT --")?;
+        if !self.fun_defs().is_empty() {
+            writeln!(f, "-- FUNCTIONS --")?;
+            for fun_def in self.fun_defs() {
+                writeln!(f, "{}", LeanFunDef(fun_def))?;
+            }
+        }
```

Comment:
Function definitions could mention constants, so we have to find a way to encode constants in a way that can be shared between definitions and theorems.